cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/KafkaAdminTopicConfigProvider.java

```diff
@@ -0,0 +1,168 @@
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import com.linkedin.kafka.cruisecontrol.config.constants.ExecutorConfig;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.Config;
+import org.apache.kafka.clients.admin.ConfigEntry;
+import org.apache.kafka.common.config.ConfigResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The Kafka topic config provider implementation based on files. The format of the file is JSON, listing properties:
+ * <pre>
+ *   {
+ *     "min.insync.replicas": 1,
+ *     "an.example.cluster.config": false
+ *   }
+ * </pre>
+ *
+ */
+public class KafkaAdminTopicConfigProvider implements TopicConfigProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaAdminTopicConfigProvider.class);
+
+  public static final String CLUSTER_CONFIGS_FILE = "cluster.configs.file";
+  private static Properties _clusterConfigs;
+  private AdminClient _adminClient;
+  private long _adminTimeoutMs;
+
+  @Override
+  //TODO: This is just returning a parsed file? Why not the cluster config from the actual cluster?
+  public Properties clusterConfigs() {
+    return _clusterConfigs;
+  }
+
+  @Override
+  public Properties topicConfigs(String topic) {
+    Config topicConfig = null;
+    ConfigResource topicResource = new ConfigResource(ConfigResource.Type.TOPIC, topic);
+    try {
+      LOG.debug("Requesting details for topic '{}'", topic);
+      topicConfig = _adminClient
+              .describeConfigs(Collections.singletonList(topicResource))
+              .all()
+              .get(_adminTimeoutMs, TimeUnit.MILLISECONDS)
+              .get(topicResource);
+    } catch (InterruptedException e) {
+      LOG.error("The request for the configuration of topic '{}' was interrupted", topic);
+      e.printStackTrace();
+    } catch (ExecutionException e) {
+      LOG.error("The request for the configuration of topic '{}' failed", topic);
+      e.printStackTrace();
+    } catch (TimeoutException e) {
+      LOG.error("The request for the configuration of topic '{}' timed out", topic);
+      e.printStackTrace();
+    }
+
+    if (topicConfig != null) {
+      return convertTopicConfigToProperties(topicConfig);
+    } else {
+      LOG.error("The configuration for topic '{}' could not be retrieved", topic);
+      return new Properties();
+    }
+  }
+
+  @Override
+  public Map<String, Properties> allTopicConfigs() {
+    Map<ConfigResource, Config> topicConfigs = null;
+    try {
+      LOG.debug("Requesting configurations for all topics");
+      topicConfigs = _adminClient
+              .listTopics()
+              .names()
```

Comment:
It is unfortunate that Kafka does not provide a direct API to retrieve all topic configs.
Given this change, we double the number of adminClient requests and roundtrip time -- i.e. once to receive topic names and another time to get their configs.

The following item doesn't have to be in this patch, but I think it should definitely be before switching to this class as the default topic config provider:
* Can we add a new method to `TopicConfigProvider` interface with the following signature?
```
Map<String, Properties> topicConfigs(Set<String> topics);
```

Adding this method would let us retrieve given topics with a single network roundtrip -- i.e. single Admin Client request. This new method would be useful to retrieve configs efficiently in `KafkaClusterState`, where we know the topics beforehand.

Reply:
I agree this would be a good thing to add but was leaving this until you are happy with the `KafkaAdminTopicProvider` implementation.

Reply:
I am implementing this change. However, I had a question about what these methods should return if an error is encountered. Under the Zookeeper client implementations the `kafka.zk.KafkaZkClient#getEntityConfigs` method will return an empty `Properties` instance if the ZK node doesn't exist and throw an exception for any other case. 

So I was wondering what the expected behavior would be for the Kafka Admin Client version. Should I throw an exception if any error is encountered or are their cases where I should return an empty `Map`/`Properties`?

Reply:
I added the multi-topic fetch method to the interface and made it so that empty properties/maps are returned in the case of most errors. Let me know what you think.